sourcecode/hub/app/Http/Controllers/GoogleController.php

```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Models\User;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Hash;
+use Laravel\Socialite\Facades\Socialite;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+
+class GoogleController extends Controller
+{
+    public function loginWithGoogle(): RedirectResponse
+    {
+        return Socialite::driver('google')->redirect();
+    }
+
+    public function callbackFromGoogle(): RedirectResponse
+    {
+        try {
+            $user = Socialite::driver('google')->user();
+            $existingUser = User::where('email', $user->getEmail())->first();
+
+            if (!$existingUser) {
+                $newUser = User::create([
+                    'name' => $user->getName(),
+                    'email' => $user->getEmail(),
+                    'password' => Hash::make($user->getName() . '@' . $user->getId())
```

Comment:
The password should be made nullable, and not be set here. In fact, doing it this way makes the password known to anyone who has the name and ID of the Google account, which is a huge security vulnerability.